### PR TITLE
feat: Ability to Manage Codespaces Secrets

### LINKED ISSUE #1662: [FEAT]: Ability to Manage Codespaces Secrets
### Describe the need

# Ability to Manage Codespaces Secrets

#915

##  Description

This feature request aims to add support for managing GitHub Codespaces secrets using the Terraform GitHub provider. Currently, the provider does not have a resource to manage secrets specifically for Codespaces. The proposed new resource, github_codespaces_*_secret, will enable users to create, update, and delete Codespace secrets through Terraform.

##  Use Case

Codespaces are widely used for development, and being able to manage their secrets with Terraform can help improve automation and management of infrastructure as code (IaC) for development environments. This feature will allow users to maintain consistency between their Codespaces and other resources managed by Terraform.

# 

### SDK Version

_No response_

### API Version

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by KenSpur**
Resolves #1662 #915

----

### Before the change?

* No Ability to Manage Codespaces Secrets

### After the change?

* Ability to Manage Codespaces Secrets using:
  * resource `github_codespaces_secret` & data `github_codespaces_secrets`, `github_codespaces_public_key`
  * resource `github_codespaces_organization_secret` & data `github_codespaces_organization_secrets`, `github_codespaces_organization_public_key`
  * resource `github_codespaces_user_secret` & data `github_codespaces_user_secrets`, `github_codespaces_user_public_key`

### Other information

*  Currently blocked by the implementation of the 'go-github' client [go-github/issues/2362](https://github.com/google/go-github/issues/2362) which is getting there.

----

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

### Pull request type

`Type: Feature`

### CODE REVIEW

**Comment by KenSpur** [comment]
The 'go-github' client still needs to be updated once its ready

**Comment by KenSpur** [comment]
Should i add the org and user secrets in this PR aswel or better make new once for those?

**Comment by kfcampbell** [comment]
@KenSpur thank you for contributing! Do you mind adding the corresponding documentation and link to documentation in the `github.erb` file for these new resources and data sources?

**Comment by KenSpur** [comment]
@kfcampbell Yes will do, also looking into testing etc.

**Comment by KenSpur** [comment]
@kfcampbell 

* repository functionality is manualy tested and works (using terraform, example method).
* user functionality is manualy tested and works (using terraform, example method).
* organization functionality currently getting 404's I am guesing because of org settings / my token  

concerning automated tests, Ill need to take more time to look into them better

**Comment by kfcampbell** [comment]
@KenSpur thank you for contributing? Is this ready for review/merging?

**Comment by KenSpur** [comment]
@kfcampbell I was able to manualy test the 'github_codespaces_secret' and 'github_codespaces_secret'.

Concerning  'github_codespaces_organization_secret' i am getting 404's which might be because of my token and org config, since when i uses the api or cli i get the same. 

Automated testing still needs a look which ill do today.

Also i have found a bug in how the repo secrets handle an update to the repoName, they should probably forceNew as its used in the resource id which can handle changes if im correct. I am wondering if i should fix this in this PR or keep it similar to the other secrets and fix it later?

Il update my progress later today.

**Comment by KenSpur** [comment]
Ah found the issue concerning the 404's on 'github_codespaces_organization_secret', i needed to upgrade my account to the 4$ plan.

It is weird though and might be a bug in the api because manualy trough the webapp i was able to create a codespace secret on the non upgraded org on 06/21 (2days ago).

**Comment by KenSpur** [comment]
@kfcampbell ready for review/merging!

Ill open a bug Issue concerning repository renames, so that all repo secrets can be fixed at once in a focused manner ^^

File: github/resource_github_codespaces_secret.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceGithubCodespacesSecret() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGithubCodespacesSecretCreateOrUpdate,
+		Read:   resourceGithubCodespacesSecretRead,
+		Update: resourceGithubCodespacesSecretCreateOrUpdate,
+		Delete: resourceGithubCodespacesSecretDelete,
+
```
Comment (kfcampbell) [inline github/resource_github_codespaces_secret.go:19]: Why does this resource not support importing?

File: website/docs/r/codespaces_secret.html.markdown [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+* `repository`      - (Required) Name of the repository
+* `secret_name`     - (Required) Name of the secret
+* `encrypted_value` - (Optional) Encrypted value of the secret using the GitHub public key in Base64 format.
+* `plaintext_value` - (Optional) Plaintext value of the secret to be encrypted
+
+## Attributes Reference
+
+* `created_at`      - Date of actions_secret creation.
+* `updated_at`      - Date of actions_secret update.
```
Comment (kfcampbell) [inline website/docs/r/codespaces_secret.html.markdown:54]: If/when import is added for codespaces secrets, we should update this doc here.

**Review by kfcampbell — COMMENTED** [review]
I had a question around importation, otherwise this looks good to me.

File: github/resource_github_codespaces_secret.go (hunk 1 above)
Comment (KenSpur) [reply github/resource_github_codespaces_secret.go:19]: It seems to not be implemented on the `resourceGithubActionsSecret` or `resourceGithubDependabotSecret` aswel, though org secrets do have it. Should i open an issue to do all 3?

Comment (kfcampbell) [reply github/resource_github_codespaces_secret.go:19]: @KenSpur that would be excellent, thanks! In an ideal world, each resource should support importing.

**Review by kfcampbell — APPROVED** [review]
I'm happy to have importing tackled in separate PRs. Thank you for contributing!

**Pull request merged**